perftest/PerfTest.cpp

```diff
@@ -0,0 +1,275 @@
+    if (found) {
+        LOG_DBG("PerfTest waitForMessage Found message: " + substring + " after " + std::to_string(seconds) + "s");
+    } else {
+        abort("Timed out waiting for message \"" + substring + "\" after " + std::to_string(seconds) + "s");
+    }
+}
+
+void PerfTest::waitForIdle(size_t timeout)
+{
+    LOG_TRC("PerfTest waitForIdle");
+    sendMessage("uno .uno:WaitForIdle");
+    waitForMessage("WaitForIdle",timeout);
+}
+
+void PerfTest::sendMessage(const std::string &message)
+{
+    LOG_TRC("PerfTest sendMessage: " + message);
+    _handler->sendMessage(message);
+}
+
+void PerfTest::sleep(unsigned int millis)
+{
+    LOG_DBG("PerfTest sleep: " + std::to_string(millis) + "ms");
+    std::this_thread::sleep_for(std::chrono::milliseconds(millis));
+}
+
+void PerfTest::disconnect()
+{
+    LOG_DBG("PerfTest disconnect");
+    _handler->shutdown();
+}
+
+CyclePerfTest::CyclePerfTest(const std::string &name, const std::string &server) :
+    PerfTest(name, server)
+{
+}
+
+void CyclePerfTest::startMeasurement()
+{
+    LOG_DBG("CyclePerfTest startMeasurement");
+    PerfTest::startMeasurement();
+    pid_t pid;
+    pid = vfork();
+    if (pid < 0) {
+        perror("fork");
+        Util::forcedExit(EX_SOFTWARE);
+    } else if (pid == 0) {
+        LOG_DBG("Starting perf");
+        std::string pid_str = "--pid="+std::to_string(getCoolwsdPid());
+        std::string output_str = "--output="+_fileName+".data";
+        LOG_DBG("Starting perf"<<"perf"<<"record"<<"-s"<<"-e"<<"cycles"<<"--freq=1000"<<"--call-graph"<<"dwarf"<<pid_str.c_str()<<output_str.c_str());
+        execlp("perf","perf","record","-s","-e","cycles","--freq=1000","--call-graph","dwarf",pid_str.c_str(),output_str.c_str(),(char *)NULL);
+    } else {
+        child_pid = pid;
+    }
+}
+
+void CyclePerfTest::stopMeasurement()
+{
+    LOG_DBG("CyclePerfTest stopMeasurement");
+
+    // Stop measurement
+    PerfTest::stopMeasurement();
+    kill(child_pid,SIGINT);
+    waitpid(child_pid, nullptr, 0);
+    addResult("dataFile",_fileName+".data");
+
+    // Get result
+    // Use perf report to get summary statistics from the data file
+    // Use grep and awk to get the cpu cycle counts from the report output
+    // Use paste to format list of numbers for bc
+    // Use bc to sum numbers. bc, as an arbitrary precision calculator, is
+    // necessary because cycle counts can be >2^31 (too big for awk's sum)
+    std::string command = "perf report -i "+_fileName+".data --no-inline | grep \"# Event count (approx.): \" | awk '{print $5}' | paste -s -d+ | bc";
+    std::string cycles = getStringPopen(command);
+    addResult("cycles",cycles);
+
+    // Create flamegraph
+    std::string flamegraphCommand = "perf script -i "+_fileName+".data --no-inline | stackcollapse-perf.pl | flamegraph.pl > "+_fileName+".svg";
+    int systemResult = system(flamegraphCommand.c_str());
```

Comment:
Just to comment on point 3): much better, now I think only this one remains for system(). I think you could replace this one with a combination of pipe(), fork(), dup2(), close() and execvp() or something similar. Could you please look at that? bash is terrible language, let's not introduce it here (think of escaping special filenames, etc). Thanks. :-)